xymonnet/xymonnet2.1
.TH XYMONNET2 1 "Version 4.3.12: 24 Jul 2013" "Xymon"
.SH NAME
xymonnet2 \- Xymon network test master
.SH SYNOPSIS
.B "xymonnet2 [options]
.br
(See the OPTIONS section for a description of the available command-line options).

.SH DESCRIPTION
.I xymonnet2(1)
is responsible for scheduling the Xymon network tests. It also performs
some of the network tests (most TCP-based tests), but will hand off
certain types of tests to other modules via the
.I netmodule(1)
utility.

Unlike previous versions, xymonnet2 is designed to run as a permanent daemon.

xymonnet2 has built-in support for testing most plain-text and
SSL-encrypted TCP-based protocols. It also supports the DNS and NTP
protocols, which are UDP-based. The full list of tests is found in the
.I protocols2.cfg(5)
file in $XYMONHOME/etc/protocols2.cfg. You can add more network tests,
as long as they are TCP-based and can be adequately tested via the
scripting language in the protocols2.cfg file.

Currently, "conn" (ping), "rpc" and "ldap" tests handled by add-on
modules. xymonnet2 will schedule these tests for the "netmodule" utility,
which is then responsible for performing the tests.

.SH HOSTNAME RESOLUTION
xymonnet2 performs the connectivity test (ping) based on the
hostname, unless the host is tagged with "testip" or the 
"--dns=ip" option is used. So the target of the connectivity
test can be determined by your /etc/hosts file or DNS.

Since xymonnet2 performs so many DNS lookups that it can severely
flood a DNS server, the results are cached for 1 hour between tests. 
This is currently a fixed value which disregards any TTL settings
in the DNS record.


.SH GENERAL OPTIONS

.IP --timeout=N
Determines the default timeout (in seconds) for each network test.
Individual tests can be given a specific timeout via the "nopt" 
setting in 
.I hosts.cfg(5)
.br
Default: 30 seconds

.IP --concurrency=N 
Determines the number of network tests that
run in parallel. Default is operating system dependent,
but will usually be 256. If xymonnet begins to complain 
about not being able to get a "socket", try running
xymonnet2 with a lower value like 50 or 100.

.IP --dns=[ip|only|standard]
Determines how xymonnet2 finds the IP adresses of the hosts to test. 
By default (the "standard"), xymonnet2 does a DNS lookup of the hostname 
to determine the IP address, unless the host has the "testip" 
tag, or the DNS lookup fails.
.br
With "--dns=only" xymonnet2 will ONLY do the DNS lookup;
if it fails, then all services on that host will be 
reported as being down.
.br
With "--dns=ip" xymonnet2 will never do a DNS lookup;
it will use the IP adresse specified in hosts.cfg for
the tests. Thus, this setting is equivalent to having
the "testip" tag on all hosts.

.IP --test-untagged
When using the XYMONNETWORK environment variable to test
only hosts on a particular network segment, xymonnet2
will ignore hosts that do not have any "NET:x" tag.
So only hosts that have a NET:$XYMONNETWORK tag will be 
tested.
.br
With this option, hosts with no NET: tag are included
in the test, so that all hosts that either have a
matching NET: tag, or no NET: tag at all are tested.

.IP --source-ip=IPADDRESS
On multi-homed hosts, this option can be used to explicitly
select the source IP address used for the network tests.
"IPADDRESS" must be a valid IP-address on the host running
xymonnet.
.br
Note: This option may be listed twice, once for the default
IPv4 source address, and once for the default IPv6 source address.

.IP --wipe-db
Clear all data from the xymonnet2 test database. May be used to
ensure your tests start from scratch (no cached DNS records, no
record of when tests last ran etc.) Note that xymonnet2 terminates
immediately after clearing the database.


.IP --no-ping
Disable the connectivity test.


.SH DEBUGGING OPTIONS
.IP --no-update
Don't send any status updates to the Xymon server. Instead,
all messages are dumped to stdout.

.IP --debug
Dumps a bunch of status about the tests as they
progress to stdout. Note that this can be very verbose.


.SH INFORMATIONAL OPTIONS
.IP "--help or -?"
Provide a summary of available command-line options.

.IP "--version"
Prints the version number of xymonnet


.SH ABOUT SSL CERTIFICATE CHECKS
When xymonnet tests services that use SSL- or TLS-based protocols,
it will check that the server certificate has not expired. This check
happens automatically for https (secure web), pop3s, imaps, nntps
and all other SSL-enabled services (except ldap, see LDAP TESTS 
below).

All certificates found for a host are reported in one status message.

Note: On most systems, the end-date of the certificate is limited to
Jan 19th, 2038. If your certificate is valid after this date, xymonnet
will report it as valid only until Jan 19, 2038. This is due to
limitations in your operating system C library. See
http://en.wikipedia.org/wiki/2038_problem .


.SH ENVIRONMENT VARIABLES
.IP XYMONNETWORK
Defines the network segment where xymonnet is currently running.
This is used to filter out only the entries in the
.I hosts.cfg(5)
file that have a matching "NET:LOCATION" tag, and execute the
tests for only those hosts.

.SH FILES
.IP "~/server/etc/protocols2.cfg"
This file contains definitions of TCP services that xymonnet
can test. Definitions for a default set of common services is built
into xymonnet, but these can be overridden or supplemented by
defining services in the protocols.cfg file. See 
.I protocols2.cfg(5)
for details on this file.

.SH "SEE ALSO"
hosts.cfg(5), protocols2.cfg(5), xymonserver.cfg(5)



Changes
Changes from 4.3.11 -> 4.3.12 (24 Jul 2013)
===========================================
* rev 7211

* Security fix: Guard against directory traversal via hostname in "drophost" commands
* Fix crash in xymongen introduced in 4.3.11
* SCO client: Fix overflow in memory calculation when >2 GB memory
* Fix so "include" and "directory" definitions in configuration files now handle <tab> after the keyword
* Fix for the Xymon webpage menu on iPad's and Android (touch devices)
* Fix "drophost" handling so the host data directory is also cleared
* xymond_rrd now processes data from "clear" status messages
* Xymon clients now report the version number in the client data
* Linux clients now align "ps" output so it is more readable.
* New "generic" client message handler allows log/file monitoring from systems that are not known to Xymon.
* The Xymon client now works if invoked with a relative path to the runclient.sh script
* Other minor / internal bugfixes

Changes from 4.3.10 -> 4.3.11 (21 Apr 2013)
===========================================
* rev 7188

* Fix wrong file permissions when installing
* Linux client: Fix handling of root filesystem when mounted on "/dev/root"
* trends webpage: Fix case where hostname disappears after zoom.
* FreeBSD client: Memory patch for FreeBSD 8.0+
* xymond_alert: Fix problem with UNMATCHED rules triggering when there are
  actual recipients, but their alerts are suppressed due to a REPEAT
  setting not having expired.
* xymond_rrd: Dont crash if called with an empty status/data message
* xymond_channel: Report cause when channel-child exits/crashes
* xymongen: Geneate an overview page with only reds (like non-green)
* xymongen: Optionally define env. variable BOARDFILTER to select
  hosts/tests included in the generated pages
* links: Add pdf, docx and odt as known document formats
* Fix potential crashes after an alert cookie expired
* Fix potential crash after deleting/renaming a host
* Speedup loading of the hosts.cfg file, noticeable with very
  large hosts.cfg files (100.000+ hosts)

Changes from 4.3.9 -> 4.3.10 (6 Aug 2012)
=========================================
* rev 7164

* Fix build problems with "errno"
* Fix build problems with OpenSSL in non-default locations
* Fix build problems with certain LDAP configurations
* Fix build problems with RRDtool on FreeBSD / OpenBSD
* Fix problem with ifstat data from Fedora in graphs
* "inode" check on FreeBSD, OpenBSD, OSX, Solaris, HP/UX, AIX
  in addition to existing support for Linux
* Document building and installing Xymon on common platforms 
  (Linux, FreeBSD, OpenBSD, Solaris)
* Enhance xymoncfg so it can be used to import Xymon configuration
  settings into shell-scripts.

Changes from 4.3.8 -> 4.3.9 (15 Jul 2012)
=========================================
* rev 7120

* Fix crash when XYMSRV is undefined but XYMSERVERS is
* Fix error in calculating combo-status messages with
  forward references
* Fix error in disable-until-TIME or disable-until-OK code
* Fix documentation of DURATION in alerts.cfg / xymond_alert so
  it is consistenly listed as being in "minutes". 
* Permit explicit use of ">" and ">=" in alerts.cfg
* Permit building without the RRDtool libraries, e.g. for
  a network-tester build, but with trend-graphing disabled.
* Full compiler-warning cleanup
* Various configuration/build-script issues fixed.

Changes from 4.3.7 -> 4.3.8 (15 Jul 2012)
=========================================
* rev 7082

Bugfixes
* Workaround for DNS timeout handling, now fixed at approximately 25
  seconds.
* "hostinfo" command for xymond documented
* confreport only shows processes that are monitored
* analysis.cfg parsing of COLOR for UP rules was broken
* RRD handlers no longer crash after receiving 1 billion updates
* Using .netrc for authentication could crash xymonnet
* "directory" includes would report the wrong filename for missing
  directories.
* useradm CGI would invoke htpassword twice
* "include" and "directory" now ignores trailing whitespace
* SSLv2 support disabled if SSL-library does not support it
* Minor bugfixes and cleanups of compiler warnings.

Enhancements
* Service status on info page now links to the detailed status page.
* Add RRDGRAPHOPTS setting to permit global user-specified RRD options,
  e.g. for font to showgraph CGI
* Add check for the size of public keys used in SSL certificates
  (enabled via --sslkeysize=N option for xymonnet)
* Optionally disable the display of SSL ciphers in the sslcert status
  (the --no-cipherlist option for xymonnet)
* Improved build-scripts works on newer systems with libraries in
  new and surprising places
* Reduce xymonnet memory usage and runtime for ping tests when there
  are multiple hosts.cfg entries with the same IP-address.
* Add code for inode-monitoring on Linux. Does not currently work on
  any other client platform.
* Added the ability to disable tests until a specific time, instead of
  for some interval. Disabling a test also now computes the expire time
  for the disable to happen at the next closest minute.


Changes from 4.3.6 -> 4.3.7 (13 Dec 2011)
=========================================
* rev 6803
* Fix acknowledge CGI (broken in 4.3.6)
* Fix broken uptime calculation for systems reporting "1 day"
* Workaround Solaris breakage in the LFS-support detection
* Fix/add links to the HTML man-page index.
* Fix "Stop after" value not being shown on the "info" page.
* Fix broken alert texts when using FORMAT=SMS
* Fix wrong description of xymondboard CRITERIA in xymon(1)
* Fix missing columnname in analysis.cfg(5) DS example
* Fix missing space in output from disk IGNORE rules in
  xymond_client --dump-config
* Fix overwrite of xymon-apache.conf when upgrading
* Fix installation so it does not remove include/directory 
  lines from configuration files.
* Add client/local/ directory for custom client script

Changes from 4.3.5 -> 4.3.6 (5 Dec 2011)
========================================
* rev 6788

xymonnet/xymonnet2.c
/*----------------------------------------------------------------------------*/
/* Xymon monitor network test tool.                                           */
/*                                                                            */
/* Copyright (C) 2011-2012 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <string.h>
#include <stdio.h>
#include <signal.h>

#include "libxymon.h"

#include "setuptests.h"
#include "tcptalk.h"
#include "netdialog.h"
#include "sendresults.h"
#include "netsql.h"

#define DEF_TIMEOUT 30

time_t lastloadtime = 0;
int running = 1;
char *location = NULL;

int concurrency = 0;
int defaulttimeout = DEF_TIMEOUT;
char *defaultsourceip4 = NULL, *defaultsourceip6 = NULL;
int pingenabled = 1;
int usebackfeedqueue = 0;

void sig_handler(int signum)
{
	switch (signum) {
	  case SIGHUP:
		lastloadtime = 0;
		break;
	  case SIGINT:
	  case SIGTERM:
		running = 0;
	}
}


int run_tests(void)
{
	int count;
	listhead_t *resulthead = NULL;

	count = setup_tests_from_database(pingenabled, (running == 0));
	if (count > 0) {
		resulthead = run_net_tests(concurrency, defaultsourceip4, defaultsourceip6);
		count = resulthead->len;
		send_test_results(resulthead, programname, 0, location, usebackfeedqueue);
		cleanup_myconn_list(resulthead);
	}

	return count;
}


int main(int argc, char **argv)
{
	int argi;
	time_t nextrun = 0;

	libxymon_init(argv[0]);
	for (argi=1; (argi < argc); argi++) {
		if (standardoption(argv[argi])) {
			if (showhelp) return 0;
		}
		else if (argnmatch(argv[argi], "--timeout=")) {
			char *p = strchr(argv[argi], '=');
			defaulttimeout = atoi(p+1);
			if (defaulttimeout <= 0) defaulttimeout = DEF_TIMEOUT;
		}
		else if (argnmatch(argv[argi], "--concurrency=")) {
			char *p = strchr(argv[argi], '=');
			concurrency = atoi(p+1);
			if (concurrency <= 0) concurrency = 0;
		}
		else if (strcmp(argv[argi], "--test-untagged") == 0) {
			test_nonet_hosts(1);
		}
		else if (argnmatch(argv[argi], "--dns=")) {
			char *p = strchr(argv[argi], '=');
			if (strcmp(p+1, "ip") == 0) set_dns_strategy(DNS_STRATEGY_IP);
			else if (strcmp(p+1, "only") == 0) set_dns_strategy(DNS_STRATEGY_HOSTNAME);
			else set_dns_strategy(DNS_STRATEGY_STANDARD);
		}
		else if (argnmatch(argv[argi], "--source-ip=")) {
			char *p = strchr(argv[argi], '=');
			switch (conn_is_ip(p+1)) {
			  case 4: defaultsourceip4 = p; break;
			  case 6: defaultsourceip6 = p; break;
			}
		}
		else if ((strcmp(argv[argi], "--noping") == 0) || (strcmp(argv[argi], "--no-ping") == 0)) {
			pingenabled = 0;
		}
		else if (strcmp(argv[argi], "--once") == 0) {
			running = 0;
		}
		else if (strcmp(argv[argi], "--net-debug") == 0) {
			conn_register_infohandler(NULL, 7);
		}
		else if ((strcmp(argv[argi], "--wipedb") == 0) || (strcmp(argv[argi], "--wipe-db") == 0)) {
			xymon_sqldb_flushall();
			errprintf("Xymon net-test database wiped\n");
			return 0;
		}
		else if (*(argv[argi]) != '-') {
			add_wanted_host(argv[argi]);
			running = 0;	/* When testing specific hosts, assume "--once" */
		}
	}


	{
		struct sigaction sa;

		setup_signalhandler(programname);
		memset(&sa, 0, sizeof(sa));
		sa.sa_handler = sig_handler;
		sigaction(SIGTERM, &sa, NULL);
		sigaction(SIGINT, &sa, NULL);
		sigaction(SIGHUP, &sa, NULL);
	}

	if (xymon_sqldb_init() != 0) {
		errprintf("Cannot open Xymon SQLite database - aborting\n");
		return 1;
	}

	init_tcp_testmodule();
	usebackfeedqueue = (sendmessage_init_local() > 0);

	/* See what network we'll test */
	location = xgetenv("XYMONNETWORK");
	if (strlen(location) == 0) location = NULL;

	do {
		int testcount;
		time_t now = gettimer();

		if (now > (lastloadtime+600)) {
			lastloadtime = now;
			read_tests_from_hostscfg(defaulttimeout);
			load_protocols(NULL);
		}

		dbgprintf("Launching tests\n");
		testcount = run_tests();
		dbgprintf("Ran %d tests\n", testcount);

		if (running) {
			int timetonext;

			xymon_sqldb_sanitycheck();
			timetonext = xymon_sqldb_secs_to_next_test();
			if (timetonext > 0) {
				timetonext++;
				if (timetonext > 60) timetonext = 60;
				dbgprintf("Sleeping %d seconds\n", timetonext);
				sleep(timetonext);
			}
		}
	} while (running);

	if (usebackfeedqueue) sendmessage_finish_local();
	conn_deinit();
	xymon_sqldb_shutdown();

	return 0;
}

